src/main/java/ai/promoted/delivery/client/InsertionFactory.java

```diff
@@ -16,10 +19,50 @@ public class InsertionFactory {
    * @return a populated Insertion
    */
   public static Insertion createInsertionWithProperties(String contentId, Map<String, Object> properties) {
-    Properties props = new Properties().structField(properties);
-    return new Insertion().contentId(contentId).properties(props);
+    Insertion ins = new Insertion().contentId(contentId);
+    if (!properties.isEmpty()) {
+      ins.properties(Properties.newBuilder().setStruct(convertMapToStruct(properties)).build());
+    }
+    return ins;
+  }
+
+  private static com.google.protobuf.Struct convertMapToStruct(Map<String, Object> map) {
+    com.google.protobuf.Struct.Builder structBuilder = com.google.protobuf.Struct.newBuilder();
+    map.forEach((key, value) -> {
+      com.google.protobuf.Value protoValue = convertObjectToValue(value);
+      structBuilder.putFields(key, protoValue);
+    });
+    return structBuilder.build();
   }
-  
+
+  private static com.google.protobuf.Value convertObjectToValue(Object object) {
+    if (object instanceof Boolean) {
+      return com.google.protobuf.util.Values.of((Boolean) object);
+    } else if (object instanceof Integer) {
+      return com.google.protobuf.util.Values.of((Integer) object);
+    } else if (object instanceof Long) {
+      return com.google.protobuf.util.Values.of((Long) object);
+    } else if (object instanceof Float) {
+      return com.google.protobuf.util.Values.of((Float) object);
+    } else if (object instanceof Double) {
+      return com.google.protobuf.util.Values.of((Double) object);
+    } else if (object instanceof String) {
+      return com.google.protobuf.util.Values.of((String) object);
+    } else if (object instanceof Iterable) {
+      Iterable<?> iterable = (Iterable<?>) object;
+      ListValue.Builder listBuilder = ListValue.newBuilder();
+      for (Object item : iterable) {
+        listBuilder.addValues(convertObjectToValue(item));
+      }
+      return com.google.protobuf.util.Values.of(listBuilder.build());
+    } else if (object instanceof Map) {
```

Comment:
I'd put Map above Iterable.  I know Map doesn't extend Iterable but it made me do a double take.

Reply:
Done.